### PR TITLE
Replace ActiveRecord dynamic finders

### PR DESCRIPTION
**Pull request by spaghetticode**
See issue #1881

### CODE REVIEW

File: backend/spec/controllers/spree/admin/orders/customer_details_controller_spec.rb [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
     let(:order) { create(:order, number: "R123456789") }
 
-    before { allow(Spree::Order).to receive(:find_by_number!) { order } }
+    before { allow_any_instance_of(Spree::Order::ActiveRecord_Relation).to receive(:find_by!) { order } }
```
Comment (spaghetticode) [inline backend/spec/controllers/spree/admin/orders/customer_details_controller_spec.rb]: I had to change the code as shown or the specs would fail. 

The reason behind this is unclear to me, as I didn't expect that changing only the message name (from `:find_by_number!` to `:find_by!`) would make the specs to fail, but that's what happens.

The relevant line for that `before` block is:
```
@order = Spree::Order.includes(:adjustments).find_by_number!(number: params[:order_id])
```

`Spree::Order.includes(:adjustments)` is not an instance of `Spree::Order`, but an instance of `Spree::Order::ActiveRecord_Relation`, so I am puzzled about how it worked before my changes.

Comment (tvdeyen) [inline backend/spec/controllers/spree/admin/orders/customer_details_controller_spec.rb]: A `receive_message_chain(:adjustments, :find_by!)` should do the trick here.

File: backend/spec/controllers/spree/admin/orders_controller_spec.rb [hunk 2]
```diff
@@ -30,7 +30,7 @@
     let(:adjustments) { double('adjustments') }
 
     before do
-      allow(Spree::Order).to receive_messages(find_by_number!: order)
+      allow_any_instance_of(Spree::Order::ActiveRecord_Relation).to receive_messages(find_by!: order)
       allow(order).to receive_messages(contents: Spree::OrderContents.new(order))
```
Comment (tvdeyen) [inline backend/spec/controllers/spree/admin/orders_controller_spec.rb]: We prefer using `receive_message_chain` here as we don't want to test internal active record implementation details. This specs will fail, if Rails renames it's relation class.

File: backend/app/controllers/spree/admin/option_types_controller.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ def location_after_save
       end
 
       def load_product
-        @product = Spree::Product.find_by_param!(params[:product_id])
+        @product = Spree::Product.find_by!(param: params[:product_id])
```
Comment (tvdeyen) [inline backend/app/controllers/spree/admin/option_types_controller.rb]: This method is defined by us in `core/lib/spree/core/permalinks.rb` and are included in all active record classes. This should not be replaced.

**Comment by tvdeyen** [comment]
I'm shocked that the specs passes, although we changed `find_by_param` usages 🤔

File: backend/app/controllers/spree/admin/option_types_controller.rb (hunk 3 above)
Comment (kennyadsl) [inline backend/app/controllers/spree/admin/option_types_controller.rb]: I think this is not tested, this should probably be something like: 

```ruby
find_by: Spree::Product.permalink_field.to_sym => params[:product_id]
```

**Comment by kennyadsl** [comment]
I'd remove those permalink finders methods from`core/lib/spree/core/permalinks.rb` if we are not using them or they can be replaced with `Model.permalink_field` using the default `find_by!` method.

File: backend/spec/controllers/spree/admin/orders_controller_spec.rb (hunk 2 above)
Comment (spaghetticode) [inline backend/spec/controllers/spree/admin/orders_controller_spec.rb]: Thank you for the suggestion, `receive_message_chain` did the trick indeed

File: backend/app/controllers/spree/admin/option_types_controller.rb (hunk 3 above)
Comment (spaghetticode) [inline backend/app/controllers/spree/admin/option_types_controller.rb]: I think the best thing for me, with my limited knowledge of what is happening here, is to revert the code to what it was before.

Comment (kennyadsl) [inline backend/app/controllers/spree/admin/option_types_controller.rb]: yes, it's probably better to do that in a separate PR anyway (if it is useful)

**Comment by spaghetticode** [comment]
@mamhoff That's a good point, thank you. I changed the commit description and the PR title as well with a more descriptive text.

**Comment by picazoH** [comment]
nice PR :clap: thank you @spaghetticode

**Comment by mamhoff** [comment]
Please squash both commits into one (the first one would introduce a commit at which we'd have a failing test suite, which is no good when you try `git bisect` to find which commit introduced regressions). 

Also, your commit message's title is too long (50 characters is what you should aim for), and the body of the message reads like the commit title of another commit, not like an explanation of why you did that change. I encourage reading the aforementioned blog post again! 

Like I said before, I am absolutely fine with the content of the PR. But please be considerate to future readers of this code base.

**Comment by spaghetticode** [comment]
@mamhoff Sorry I took the commit message matter too lightly. I hope it's now fine with you.

**Comment by mamhoff** [comment]
@spaghetticode The commit message is much improved now, thank you for that! 

@tvdeyen if you have a minute could you re-review this? I think your concerns have been addressed, too.
